### PR TITLE
Redirectable stdin trinket

### PR DESCRIPTION
**Pull request by albertjan**
@eah13 @brianpmarks @waywaaard I accidentally pulled in some other stuff aswell I don't think it will hurt trinket. Leave it in or cherrypick what do you say.

This PR adds the ability to redirect stdin and out. It handles stdin and out as "real" files. the sys module now has `stdin`, `__stdin__`, `stdout` and `__stdout__`. And I've added the `StringIO` module, so you have something to write to and read from.

### CODE REVIEW

**Comment by eah13** [comment]
The list of other skulpt changes looks good- don't think there are any potential breakages in there to be worried about.  improvements ro random always welcome.

Looking forward to stesting this once I'm back from the grave :syringe: :hospital:

**Comment by brianpmarks** [comment]
+1 from me to leave in the additions. We've actually got someone asking about the processing fix so it'd be good to get that deployed.

**Comment by waywaaard** [comment]
+1

**Comment by brianpmarks** [comment]
@albertjan When running `./skulpt.py dist` on this branch I got lots of complaints. Below is the first one which I think may exemplify the issue.

Are there different build instructions or is this an artifact of the tests needing to be updated for the changes?

```
FAILED: (test/run/t00.py)
-----
print "hello world"

-----
GOT:
-----
EXCEPTION: IOError: File not open for writing on line 1

-----
WANTED:
-----
hello world

-----
DIFF:
-----
len got: 56

len wanted: 12

firstdiff at: 0 got: 69 (EXCEPTION: IOError: File not open for writing on line 1
) expect: 104 (hello world
)
```

**Comment by albertjan** [comment]
Oh yeah that probably happend during the merge, will fix tonight :)

**Comment by albertjan** [comment]
@brianpmarks that should do the trick :)

**Comment by brianpmarks** [comment]
I got this friendly message: `src/file.js: line 216, col 16, Expected indentation of 16 characters`

Indenting that line by 2 extra spaces seemed to make everything happy. :smile:

**Comment by albertjan** [comment]
oh seriously jscs must not be installed on my machine

**Comment by brianpmarks** [comment]
@albertjan Looks good!

A couple of questions. :smile:

We had hacked in some file writing stuff a while back. See the relevant skulpt and trinket links below:

https://github.com/trinketapp/skulpt/blob/master/src/file.js#L47
https://github.com/trinketapp/skulpt/blob/master/src/file.js#L160

https://github.com/trinketapp/trinket/blob/develop/public/js/skulpt/wrapper.js#L497
https://github.com/trinketapp/trinket/blob/develop/public/js/skulpt/wrapper.js#L504

One change I noticed is that in the trinket `fileopen` and `filewrite` functions we were getting a skulpt object for the file name and we had to reach in and grab the `.v` attribute. That's no longer needed I think because of [this](https://github.com/trinketapp/skulpt/pull/10/files#diff-de48719f755e5f35fe85c57bf1830fc1R16). Is that right? (This seems like a good change to me. I just wanted to make sure I knew why and what to update on the trinket side.)

An unintended side-effect though is that anytime `StringIO` is imported or a file is opened trinket is getting the open file events for `/dev/stdout` and `/dev/stdin` and creating tabs for them. Any suggestions for dealing with that? We could check the filename and not create the tabs in certain situations, for example if it starts with a slash. Let me know if this isn't clear. As of now [here's an example running on my dev server](http://briandev3.trinket.io/embed/python/773b2fdb24) that shows this happening.

Lastly, any good `StringIO` examples or use cases I should test?

**Comment by albertjan** [comment]
Oh for pete's sake, sorry about this. yeah we need to filter out the /dev/ files from the fileopen event. 

the missing `.v` is because it's unpacked with `Sk.ffi.remapToJs` 

This is what I used to do testing:

```
import StringIO
import sys

inn = StringIO.StringIO("testing the stdin")
out = StringIO.StringIO()

sys.stdin = inn
sys.stdout = out

print raw_input("test")

sys.stdout = sys.__stdout__

print "out! " + out.getvalue()
```

There's an issue with writing to the stdout when you run it for the second time because something goes wrong with the /dev/ files. That should be fixed when we get rid of the tabs/files.

File: src/file.js [hunk 1]
```diff
@@ -157,18 +203,27 @@ Sk.builtin.file.prototype["truncate"] = new Sk.builtin.func(function (self, size
...
+    var mode = Sk.ffi.remapToJs(self.mode);
+    if (self.mode.v === "w" || self.mode.v === "wb" || self.mode.v === "a" || self.mode.v === "ab") {
+        if (Sk.filewrite) {
+            if (self.closed) {
+                throw new Sk.builtin.ValueError("I/O operation on closed file");
+            }
 
-        if (self.mode.v === "w" || self.mode.v === "wb" || self.mode.v === "a" || self.mode.v === "ab") {
-            Sk.filewrite(self, str);
+            if (self.fileno === 1) {
```
Comment (albertjan) [inline src/file.js]: @brianpmarks as you can see here it'll only call `Sk.filewrite` when it's not writing to the `/dev/stdout`

File: src/file.js [hunk 2]
```diff
@@ -12,39 +12,52 @@ Sk.builtin.file = function (name, mode, buffering) {
...
+        }
+        this.currentLine = 0;
     }
-    this.currentLine = 0;
     this.pos$ = 0;
 
     this.__class__ = Sk.builtin.file;
 
-    if (Sk.fileopen) {
+    if (Sk.fileopen && this.fileno >= 10) {
```
Comment (albertjan) [inline src/file.js]: this makes sure that `Sk.fileopen` is only called for real files :smile:

**Comment by brianpmarks** [comment]
Looks good!!! Thanks!

**Pull request merged**